File: akka-persistence-typed-tests/src/test/java-21+/jdocs21/akka/persistence/typed/javadsl/BlogPostEntityDurableState.java
/*
 * Copyright (C) 2018-2023 Lightbend Inc. <https://www.lightbend.com>
 */

package jdocs21.akka.persistence.typed.javadsl;

import akka.Done;
import akka.actor.typed.ActorRef;
import akka.actor.typed.Behavior;
import akka.actor.typed.javadsl.Behaviors;
import akka.persistence.typed.PersistenceId;
import akka.persistence.typed.state.javadsl.*;

// #behavior
public class BlogPostEntityDurableState
    extends DurableStateOnCommandBehavior<
        BlogPostEntityDurableState.Command, BlogPostEntityDurableState.State> {

  public sealed interface Command {}

  public record AddPost(PostContent content, ActorRef<AddPostDone> replyTo) implements Command {}
  public record GetPost(ActorRef<PostContent> replyTo) implements Command {}
  public record ChangeBody(String newBody, ActorRef<Done> replyTo) implements Command {}
  public record Publish(ActorRef<Done> replyTo) implements Command {}
  public record PostContent(String postId, String title, String body) implements Command {}
  // reply
  public record AddPostDone(String postId) {}


  public sealed interface State {}

  public record BlankState() implements State {}

  public record DraftState(PostContent content) implements State {
    DraftState withContent(PostContent newContent) {
      return new DraftState(newContent);
    }
    DraftState withBody(String newBody) {
      return withContent(new PostContent(postId(), content.title, newBody));
    }
    String postId() {
      return content.postId;
    }
  }

  public record PublishedState(PostContent content) implements State {
    PublishedState withContent(PostContent newContent) {
      return new PublishedState(newContent);
    }
    PublishedState withBody(String newBody) {
      return withContent(new PostContent(postId(), content.title, newBody));
    }
    String postId() {
      return content.postId;
    }
  }

  public static Behavior<Command> create(String entityId, PersistenceId persistenceId) {
    return Behaviors.setup(
        context -> {
          context.getLog().info("Starting BlogPostEntityDurableState {}", entityId);
          return new BlogPostEntityDurableState(persistenceId);
        });
  }

  private BlogPostEntityDurableState(PersistenceId persistenceId) {
    super(persistenceId);
  }

  @Override
  public State emptyState() {
    return new BlankState();
  }

  @Override
  public Effect<State> onCommand(State state, Command command) {
    return switch(state) {
      case BlankState ignored ->
        switch(command) {
          case AddPost addPost -> onAddPost(addPost);
          default -> Effect().unhandled();
        };
      case DraftState draft ->
        switch(command) {
          case ChangeBody changeBody -> onChangeBody(draft, changeBody);
          case Publish publish -> onPublish(draft, publish);
          case GetPost getPost -> onGetPost(draft, getPost);
          default -> Effect().unhandled();
        };
      case PublishedState published ->
        switch(command) {
          case ChangeBody changeBody -> onChangeBody(published, changeBody);
          case GetPost getPost -> onGetPost(published, getPost);
          default -> Effect().unhandled();
        };
    };
  }

  private Effect<State> onAddPost(AddPost cmd) {
    // #reply
    return Effect()
        .persist(new DraftState(cmd.content))
        .thenRun(() -> cmd.replyTo.tell(new AddPostDone(cmd.content.postId)));
    // #reply
  }

  private Effect<State> onChangeBody(DraftState state, ChangeBody cmd) {
    return Effect()
        .persist(state.withBody(cmd.newBody))
        .thenRun(() -> cmd.replyTo.tell(Done.getInstance()));
  }

  private Effect<State> onChangeBody(PublishedState state, ChangeBody cmd) {
    return Effect()
        .persist(state.withBody(cmd.newBody))
        .thenRun(() -> cmd.replyTo.tell(Done.getInstance()));
  }

  private Effect<State> onPublish(DraftState state, Publish cmd) {
    return Effect()
        .persist(new PublishedState(state.content))
        .thenRun(
            () -> {
              System.out.println("Blog post published: " + state.postId());
              cmd.replyTo.tell(Done.getInstance());
            });
  }

  private Effect<State> onGetPost(DraftState state, GetPost cmd) {
    cmd.replyTo.tell(state.content);
    return Effect().none();
  }

  private Effect<State> onGetPost(PublishedState state, GetPost cmd) {
    cmd.replyTo.tell(state.content);
    return Effect().none();
  }
}
// #behavior

File: akka-persistence-typed/src/test/scala/docs/akka/persistence/typed/BlogPostEntityDurableState.scala
/*
 * Copyright (C) 2017-2025 Lightbend Inc. <https://www.lightbend.com>
 */

package docs.akka.persistence.typed

import akka.Done
import akka.actor.typed.ActorRef
import akka.actor.typed.Behavior
import akka.actor.typed.scaladsl.Behaviors
import akka.pattern.StatusReply
import akka.persistence.typed.PersistenceId
import akka.persistence.typed.state.scaladsl.Effect
import akka.persistence.typed.state.scaladsl.DurableStateBehavior

//#behavior
object BlogPostEntityDurableState {
  // commands, state defined here

  //#behavior

  //#state
  sealed trait State

  case object BlankState extends State

  final case class DraftState(content: PostContent) extends State {
    def withBody(newBody: String): DraftState =
      copy(content = content.copy(body = newBody))

    def postId: String = content.postId
  }

  final case class PublishedState(content: PostContent) extends State {
    def postId: String = content.postId
  }
  //#state

  //#commands
  sealed trait Command
  //#reply-command
  final case class AddPost(content: PostContent, replyTo: ActorRef[StatusReply[AddPostDone]]) extends Command
  final case class AddPostDone(postId: String)
  //#reply-command
  final case class GetPost(replyTo: ActorRef[PostContent]) extends Command
  final case class ChangeBody(newBody: String, replyTo: ActorRef[Done]) extends Command
  final case class Publish(replyTo: ActorRef[Done]) extends Command
  final case class PostContent(postId: String, title: String, body: String)
  //#commands

  //#behavior
  def apply(entityId: String, persistenceId: PersistenceId): Behavior[Command] = {
    Behaviors.setup { context =>
      context.log.info("Starting BlogPostEntityDurableState {}", entityId)
      DurableStateBehavior[Command, State](persistenceId, emptyState = BlankState, commandHandler)
    }
  }
  //#behavior

  //#command-handler
  private val commandHandler: (State, Command) => Effect[State] = { (state, command) =>
    state match {

      case BlankState =>
        command match {
          case cmd: AddPost => addPost(cmd)
          case _            => Effect.unhandled
        }

      case draftState: DraftState =>
        command match {
          case cmd: ChangeBody  => changeBody(draftState, cmd)
          case Publish(replyTo) => publish(draftState, replyTo)
          case GetPost(replyTo) => getPost(draftState, replyTo)
          case AddPost(_, replyTo) =>
            Effect.unhandled[State].thenRun(_ => replyTo ! StatusReply.Error("Cannot add post while in draft state"))
        }

      case publishedState: PublishedState =>
        command match {
          case GetPost(replyTo) => getPost(publishedState, replyTo)
          case AddPost(_, replyTo) =>
            Effect.unhandled[State].thenRun(_ => replyTo ! StatusReply.Error("Cannot add post, already published"))
          case _ => Effect.unhandled
        }
    }
  }

  private def addPost(cmd: AddPost): Effect[State] = {
    //#reply
    Effect.persist(DraftState(cmd.content)).thenRun { _ =>
      // After persist is done additional side effects can be performed
      cmd.replyTo ! StatusReply.Success(AddPostDone(cmd.content.postId))
    }
    //#reply
  }

  private def changeBody(state: DraftState, cmd: ChangeBody): Effect[State] = {
    Effect.persist(state.withBody(cmd.newBody)).thenRun { _ =>
      cmd.replyTo ! Done
    }
  }

  private def publish(state: DraftState, replyTo: ActorRef[Done]): Effect[State] = {
    Effect.persist(PublishedState(state.content)).thenRun { _ =>
      println(s"Blog post ${state.postId} was published")
      replyTo ! Done
    }
  }

  private def getPost(state: DraftState, replyTo: ActorRef[PostContent]): Effect[State] = {
    replyTo ! state.content
    Effect.none
  }

  private def getPost(state: PublishedState, replyTo: ActorRef[PostContent]): Effect[State] = {
    replyTo ! state.content
    Effect.none
  }
  //#command-handler
  //#behavior

  // commandHandler defined here
}
//#behavior
